### DISCUSSION TITLE
How to pickle the `client` object

### DISCUSSION THREAD

**Discussion by HYB777** [MPh-py/MPh#180] [General]
Hi, I want to use [ray](https://github.com/ray-project/ray) for multi-nodes parallel computing on the cluster. And the ray use [cloudpickle](https://github.com/cloudpipe/cloudpickle/tree/master) to serializate the object.

But the mph object is java object, so I use JPickle for serialization. 

But it raise the following error

> Traceback (most recent call last):
  File "org.jpype.pickle.Encoder.java", line -1, in org.jpype.pickle.Encoder.pack
Exception: Java Exception
>
>The above exception was the direct cause of the following exception:
>
>Traceback (most recent call last):
  File "/home/hyb/hyb/ShapeOptDemos/shape-rl/rl-project/test_pickle.py", line 33, in <module>
    inband = pickle.dumps(
  File "/home/hyb/anaconda3/envs/hyb/lib/python3.9/site-packages/ray/cloudpickle/cloudpickle.py", line 1481, in dumps
    cp.dump(obj)
  File "/home/hyb/anaconda3/envs/hyb/lib/python3.9/site-packages/jpype/pickle.py", line 102, in reduce
    byte = bytes(self._encoder.pack(obj))
java.io.java.io.NotSerializableException: java.io.NotSerializableException: com.comsol.clientapi.impl.ModelClient

**Comment by Thrameos** [comment]
This is an issue with the jar file.  Only classes marked Serializable can be serialized.   Either the class needs to be marked appropriately for serialization or the field holding it must be transient.

**Comment by HYB777** [comment]
> This is an issue with the jar file. Only classes marked Serializable can be serialized. Either the class needs to be marked appropriately for serialization or the field holding it must be transient.

@Thrameos Thanks for your reply, I'm not very familiar with these, is there any solution for this?

**Reply by john-hen** [reply]
It means it's not possible. We don't control the JAR files, Comsol does. Plus, this is a client-server setup (on Linux), so deserializing the pickled objects that the client holds won't preserve the connection to its server. You would have to find a way to start the Comsol session similar to how it's done in the ["Multiple processes"](https://mph.readthedocs.io/en/1.2/demonstrations.html#multiple-processes) demo, if Ray somehow let's you do that.
